File: app/src/main/java/com/dragonpav/dengine/Shader.java
package com.dragonpav.dengine;

import android.opengl.GLES30;
import android.util.Log;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

public class Shader {
	public int shaderId;
	private String source;
	public Type type;
	public Shader(InputStream file, Type type) {
		this.type = type;
		try {
			BufferedReader br = new BufferedReader(new InputStreamReader(file));
			String line;
			StringBuilder builder = new StringBuilder();
			while ((line = br.readLine()) != null) {
				builder.append(line).append('\n');
			}
			source = builder.toString();
			file.close();
		} catch (IOException e) {
			e.fillInStackTrace();
		}
		switch (type) {
			case Vertex:
				shaderId = GLES30.glCreateShader(GLES30.GL_VERTEX_SHADER);
				break;
			case Fragment:
				shaderId = GLES30.glCreateShader(GLES30.GL_FRAGMENT_SHADER);
				break;
		}
		Log.i("DEngine", "Shader with type: " + type.name() + ", id: " + shaderId + " created.");
		GLES30.glShaderSource(shaderId, source);
	}
	public boolean compile() {
		GLES30.glCompileShader(shaderId);
		int[] status = new int[1];
		GLES30.glGetShaderiv(shaderId, GLES30.GL_COMPILE_STATUS, status, 0);
		if (status[0] == 1) {
			Log.i("DEngine", "Shader with id: " + shaderId + " successfully compiled.");
			return true;
		} else {
			String infoLog = GLES30.glGetShaderInfoLog(shaderId);
			Log.w("DEngine", "Failed compile shader (type: " + type.name() + ") Error log:");
			Log.w("DEngine", "------------------------");
			Log.w("DEngine", infoLog);
			delete();
			return false;
		}
	}
	public void delete() {
		GLES30.glDeleteShader(shaderId);
	}
	public enum Type {
		Vertex,
		Fragment
	}
}

File: app/src/main/cpp/Shader.cpp
#include "Shader.h"
Shader::Shader(AAsset *asset, Type type) {
    this->type = type;
    switch (type) {
        case Vertex:
            shaderId = glCreateShader(GL_VERTEX_SHADER);
            __android_log_write(ANDROID_LOG_INFO, "DEngine", ("Created shader with id: " + std::to_string(shaderId) + ", type: Vertex").c_str());
            break;
        case Fragment:
            shaderId = glCreateShader(GL_FRAGMENT_SHADER);
            __android_log_write(ANDROID_LOG_INFO, "DEngine", ("Created shader with id: " + std::to_string(shaderId) + ", type: Fragment").c_str());
            break;
    }
    size_t fileLength = AAsset_getLength(asset);
    char* content = (char*) malloc(fileLength + 1);
    AAsset_read(asset, content, fileLength);
    content[fileLength] = '\0';
    AAsset_close(asset);
    glShaderSource(shaderId, 1, &content, nullptr);
}
void Shader::compile() {
    glCompileShader(shaderId);
    GLint status;
    glGetShaderiv(shaderId, GL_COMPILE_STATUS, &status);
    if (status) {
        __android_log_write(ANDROID_LOG_INFO, "DEngine", ("Shader with id: " + std::to_string(shaderId) + " successfully compiled").c_str());
    } else {
        GLchar infoLog[512];
        glGetShaderInfoLog(shaderId, 512, NULL, infoLog);
        __android_log_write(ANDROID_LOG_ERROR, "DEngine", ("Shader with id: " + std::to_string(shaderId) + " not compiled:").c_str());
        __android_log_write(ANDROID_LOG_WARN, "DEngine", infoLog);
    }
}
void Shader::deleteShader() {
    glDeleteShader(shaderId);
}